akka-stream-tests/src/test/scala/akka/stream/scaladsl/UnfoldResourceAsyncSourceSpec.scala

```diff
@@ -0,0 +1,246 @@
+
+      sub.request(1)
+      expectMsg(Open)
+      c.expectNoMsg(200.millis)
+      createPromise.success(resource)
+
+      val chunks = manyLinesArray.toList.iterator
+
+      expectMsg(Read(resource))
+      c.expectNoMsg(200.millis)
+      readPromise.success(Option(resource.readLine()))
+      c.expectNext() should ===(chunks.next())
+
+      sub.cancel()
+      expectMsg(Close(resource))
+      resource.close()
+      closePromise.success(Done)
+    }
+
+    "close resource successfully right after open" in assertAllStagesStopped {
+      val createPromise = Promise[BufferedReader]()
+      val readPromise = Promise[Option[String]]()
+      val closePromise = Promise[Done]()
+
+      val resource = new BufferedReader(new FileReader(manyLinesFile))
+      val p = Source.unfoldResourceAsync[String, BufferedReader](() ⇒ {
+        testActor ! Open
+        createPromise.future
+      },
+        reader ⇒ {
+          testActor ! Read(reader)
+          readPromise.future
+        },
+        reader ⇒ {
+          testActor ! Close(reader)
+          closePromise.future
+        })
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      expectMsg(Open)
+      createPromise.success(resource)
+
+      sub.cancel()
+      expectMsg(Close(resource))
+      resource.close()
+      closePromise.success(Done)
+    }
+
+    "continue when Strategy is Resume and exception happened" in assertAllStagesStopped {
+      val p = Source.unfoldResourceAsync[String, BufferedReader](open,
+        reader ⇒ {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw ex else Promise.successful(Option(s)).future
+        }, close).withAttributes(supervisionStrategy(resumingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 49).foreach(i ⇒ {
+        sub.request(1)
+        c.expectNext() should ===(if (i < 10) manyLinesArray(i) else manyLinesArray(i + 10))
+      })
+      sub.request(1)
+      c.expectComplete()
+    }
+
+    "close and open stream again when Strategy is Restart" in assertAllStagesStopped {
+      val p = Source.unfoldResourceAsync[String, BufferedReader](open,
+        reader ⇒ {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw ex else Promise.successful(Option(s)).future
+        }, close).withAttributes(supervisionStrategy(restartingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 19).foreach(i ⇒ {
+        sub.request(1)
+        c.expectNext() should ===(manyLinesArray(0))
+      })
+      sub.cancel()
+    }
+
+    "work with ByteString as well" in assertAllStagesStopped {
```

Comment:
Why is this a separate testcase, how does it differ from any other type that is returned from read?